### PR TITLE
[Sparse] Sparse matrix reduction Python part

### PR DESCRIPTION
**Pull request by BarclayII**
## Description
Implemented `.sum()`, `.smin()`, `.smax()`, `.smean()` and `.prod()`.

Completes #4865.  Depends on #5012.

One question: I saw in `mock_sparse` that minimum/maximum/mean reduction is named `smin()`, `smax()` and `smean()` but in the issues there's no `s` prefix.  Which one should we take?  I personally prefer without the `s` prefix.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [x] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### CODE REVIEW

File: python/dgl/mock_sparse2/reduction.py [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+from .sparse_matrix import SparseMatrix
+
+
+def reduce(A: SparseMatrix, dim: Optional[int] = None, rtype: str = "sum"):
+    """Compute the reduction of non-zero values in sparse matrix A along
+    the given dimension :attr:`dim`.
+
+    If :attr:`dim` is None, it reduces all the elements in the sparse
+    matrix. Otherwise, it reduces on the row (``dim=0``) or column (``dim=1``)
+    dimension, producing a tensor of shape ``(A.shape[1], ) + A.val.shape[:1]``
```
Comment (frozenbugs) [inline python/dgl/mock_sparse2/reduction.py]: (A.shape[1], ) + A.val.shape[1:]

File: python/dgl/mock_sparse2/reduction.py [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+    Parameters
+    ----------
+    A : SparseMatrix
+        Sparse matrix
+    dim : int, optional
+        The dimension to reduce.  Must be either 0 (by rows) or 1 (by columns).
+
+        If not given, the reduction will happen across all non-zero entries.
+    rtype: str
+        Reduction type, one of ``['sum', 'smin', 'smax', 'smean', 'prod']``,
```
Comment (frozenbugs) [inline python/dgl/mock_sparse2/reduction.py]: if s to callout the smax is for sparse matrix, which will only operate on non-zero entries, then prod should also be sprod.

File: python/dgl/mock_sparse2/reduction.py [hunk 3]
```diff
@@ -0,0 +1,322 @@
+"""DGL sparse matrix reduce operators."""
```
Comment (mufeili) [inline python/dgl/mock_sparse2/reduction.py:1]:
```suggestion
"""DGL sparse matrix reduce operators"""
```
not a valid sentence

File: python/dgl/mock_sparse2/reduction.py [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+    """
+    return torch.ops.dgl_sparse.reduce(A.c_sparse_matrix, rtype, dim)
+
+
+# pylint: disable=W0622
+def sum(A: SparseMatrix, dim: Optional[int] = None):
+    """Compute the sum of non-zero values in sparse matrix A along
+    the given dimension :attr:`dim`.
+
+    If :attr:`dim` is None, it reduces all the elements in the sparse matrix.
```
Comment (mufeili) [inline python/dgl/mock_sparse2/reduction.py:78]: I prefer moving this paragraph to parameter description.

File: python/dgl/mock_sparse2/reduction.py [hunk 5]
```diff
@@ -0,0 +1,322 @@
...
+    Otherwise, it reduces on the row (``dim=0``) or column (``dim=1``) dimension,
+    sproducing a tensor of shape ``(A.shape[1], ) + A.val.shape[1:]`` or
```
Comment (mufeili) [inline python/dgl/mock_sparse2/reduction.py:80]: "sproducing"?

Comment (mufeili) [inline python/dgl/mock_sparse2/reduction.py:80]: Empty space here after `(A.shape[1],`, but not `(A.shape[0],)`, better be consistent

File: python/dgl/mock_sparse2/reduction.py [hunk 6]
```diff
@@ -0,0 +1,322 @@
...
+    ``(A.shape[0],) + A.val.shape[1:]``.
+
+    Parameters
+    ----------
+    dim : int, optional
+        The dimension to reduce.
```
Comment (mufeili) [inline python/dgl/mock_sparse2/reduction.py:86]: not a valid sentence for `.`

File: python/dgl/mock_sparse2/reduction.py [hunk 7]
```diff
@@ -0,0 +1,322 @@
...
+    >>> print(A.smean(0))
+    tensor([[1, 1], [0, 0], [2, 2]])
+    >>> print(A.smean(1))
+    tensor([[1, 2], [2, 1], [0, 0], [0, 0]])
+    """
+    return torch.ops.dgl_sparse.smean(A.c_sparse_matrix, dim)
+
+
+def sprod(A: SparseMatrix, dim: Optional[int] = None):
+    """Compute the sproduct of non-zero values in sparse matrix A along
```
Comment (frozenbugs) [inline python/dgl/mock_sparse2/reduction.py:268]: `Compute the sproduct of non-zero values`
just `product` should be good.

File: python/dgl/mock_sparse2/reduction.py [hunk 8]
```diff
@@ -0,0 +1,322 @@
...
+    If :attr:`dim` is None, it reduces all the elements in the sparse matrix.
+    Otherwise, it reduces on the row (``dim=0``) or column (``dim=1``) dimension,
+    sproducing a tensor of shape ``(A.shape[1], ) + A.val.shape[1:]`` or
+    ``(A.shape[0],) + A.val.shape[1:]``.
+
+    Parameters
+    ----------
+    dim : int, optional
+        The dimension to reduce.
+
```
Comment (mufeili) [inline python/dgl/mock_sparse2/reduction.py:87]: The comments above apply to other ops.

**Review by frozenbugs — APPROVED** [review]

File: python/dgl/mock_sparse2/reduction.py (hunk 5 above)
Comment (frozenbugs) [reply python/dgl/mock_sparse2/reduction.py:80]: lol, I guess it comes from `%s /prod/sprod/g`

File: python/dgl/mock_sparse2/reduction.py [hunk 9]
```diff
@@ -0,0 +1,322 @@
...
+    >>> row = torch.tensor([0, 1, 1])
+    >>> col = torch.tensor([0, 0, 2])
+    >>> val = torch.tensor([1, 1, 2])
+    >>> A = create_from_coo(row, col, val, shape=(4, 3))
+    >>> print(A.smean())
+    tensor(1)
+    >>> print(A.smean(0))
+    tensor([1, 0, 2])
+    >>> print(A.smean(1))
+    tensor([1, 1, 0, 0])
```
Comment (mufeili) [inline python/dgl/mock_sparse2/reduction.py:251]:
```suggestion
    tensor([1, 1.5, 0, 0])
```
?

File: python/dgl/mock_sparse2/reduction.py [hunk 10]
```diff
@@ -0,0 +1,322 @@
...
+
+    >>> row = torch.tensor([0, 1, 1])
+    >>> col = torch.tensor([0, 0, 2])
+    >>> val = torch.tensor([[1, 2], [2, 1], [2, 2]])
+    >>> A = create_from_coo(row, col, val, shape=(4, 3))
+    >>> print(A.smean())
+    tensor([1, 1])
+    >>> print(A.smean(0))
+    tensor([[1, 1], [0, 0], [2, 2]])
```
Comment (mufeili) [inline python/dgl/mock_sparse2/reduction.py:260]:
```suggestion
    tensor([[1.5, 1.5], [0, 0], [2, 2]])
```
?

File: python/dgl/mock_sparse2/reduction.py (hunk 7 above)
Comment (mufeili) [inline python/dgl/mock_sparse2/reduction.py:268]:
```suggestion
    """Compute the product of non-zero values in sparse matrix A along
```

File: python/dgl/mock_sparse2/reduction.py [hunk 11]
```diff
@@ -0,0 +1,322 @@
...
+    tensor([1, 0, 2])
+    >>> print(A.sprod(1))
+    tensor([1, 2, 0, 0])
+
+    >>> row = torch.tensor([0, 1, 1])
+    >>> col = torch.tensor([0, 0, 2])
+    >>> val = torch.tensor([[1, 2], [2, 1], [2, 2]])
+    >>> A = create_from_coo(row, col, val, shape=(4, 3))
+    >>> print(A.sprod())
+    tensor([1, 1])
```
Comment (mufeili) [inline python/dgl/mock_sparse2/reduction.py:308]:
```suggestion
    tensor([4, 4])
```
?

File: tests/pytorch/mock_sparse2/test_reduction.py [hunk 12]
```diff
@@ -0,0 +1,131 @@
+import operator
+import sys
+
+import backend as F
+
+import dgl.mock_sparse2 as dglsp
+import pytest
+import torch
+
```
Comment (mufeili) [inline tests/pytorch/mock_sparse2/test_reduction.py:9]: Do we need to skip the test on windows as in other test files?

File: tests/pytorch/mock_sparse2/test_reduction.py [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+        M[row[i], col[i]] += 1
+    if op == "mean":
+        A2 = A2.sum(2)
+    else:
+        A2 = getattr(A2, op)(2)
+    M = M.view(10, 15, *([1] * (val.dim() - 1)))
+    return A2, M
+
+
+@pytest.mark.parametrize("shape", [(20,), (20, 20), (20, 20, 20)])
```
Comment (mufeili) [inline tests/pytorch/mock_sparse2/test_reduction.py:55]: What's the point of having `(20, 20, 20)`?

File: tests/pytorch/mock_sparse2/test_reduction.py [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+@pytest.mark.parametrize("use_reduce", [False, True])
+def test_reduce_along(shape, dim, empty_nnz, op, use_reduce):
+    row = torch.randint(0, 10, (20,), device=F.ctx())
+    col = torch.randint(0, 15, (20,), device=F.ctx())
+    mask = (
+        torch.bincount(
+            col if dim == 0 else row, minlength=15 if dim == 0 else 10
+        )
+        == 0
+    )
```
Comment (mufeili) [inline tests/pytorch/mock_sparse2/test_reduction.py:101]: Simplify, leave comments, explain what is `10` and `15`

File: tests/pytorch/mock_sparse2/test_reduction.py [hunk 15]
```diff
@@ -0,0 +1,131 @@
+import operator
```
Comment (mufeili) [inline tests/pytorch/mock_sparse2/test_reduction.py:1]: Ideally the tests should cover the examples.

File: tests/pytorch/mock_sparse2/test_reduction.py [hunk 16]
```diff
@@ -0,0 +1,131 @@
...
+
+    head = torch.randn(*output.shape).to(val) if output.dim() > 0 else None
+    output.backward(head)
+    output2.backward(head)
+    assert (val.grad - val2.grad).abs().max() < 1e-4
+
+
+@pytest.mark.parametrize("shape", [(20,), (20, 20), (20, 20, 20)])
+@pytest.mark.parametrize("dim", [0, 1])
+@pytest.mark.parametrize("empty_nnz", [False, True])
```
Comment (mufeili) [inline tests/pytorch/mock_sparse2/test_reduction.py:90]: Explain `empty_nnz`.

File: python/dgl/mock_sparse2/reduction.py [hunk 17]
```diff
@@ -0,0 +1,322 @@
+"""DGL sparse matrix reduce operators."""
+from typing import Optional
+
+import torch
+
+from .sparse_matrix import SparseMatrix
+
+
+def reduce(A: SparseMatrix, dim: Optional[int] = None, rtype: str = "sum"):
```
Comment (mufeili) [inline python/dgl/mock_sparse2/reduction.py:9]: Is it necessary to provide `reduce` other than the specific ops?

File: python/dgl/mock_sparse2/reduction.py [hunk 18]
```diff
@@ -0,0 +1,322 @@
...
+
+    Parameters
+    ----------
+    A : SparseMatrix
+        Sparse matrix
+    dim : int, optional
+        The dimension to reduce.  Must be either 0 (by rows) or 1 (by columns).
+
+        If not given, the reduction will happen across all non-zero entries.
+    rtype: str
```
Comment (mufeili) [inline python/dgl/mock_sparse2/reduction.py:29]:
```suggestion
    rtype: str, optional
```

File: python/dgl/mock_sparse2/reduction.py [hunk 19]
```diff
@@ -0,0 +1,322 @@
...
+    >>> val = torch.tensor([[1, 2], [2, 1], [2, 2]])
+    >>> A = create_from_coo(row, col, val, shape=(4, 3))
+    >>> print(A.reduce(reduce='sum'))
+    tensor([5, 5])
+    >>> print(A.reduce(0, 'sum'))
+    tensor([[3, 3], [0, 0], [2, 2]])
+    >>> print(A.reduce(1, 'smin'))
+    tensor([[1, 2], [2, 1], [0, 0], [0, 0]])
+    >>> print(A.reduce(0, 'smean'))
+    tensor([[1, 1], [0, 0], [2, 2]])
```
Comment (mufeili) [inline python/dgl/mock_sparse2/reduction.py:68]:
```suggestion
    tensor([[1.5, 1.5], [0, 0], [2, 2]])
```

**Review by mufeili — CHANGES_REQUESTED** [review]
done a pass.

File: tests/pytorch/mock_sparse2/test_reduction.py (hunk 12 above)
Comment (BarclayII) [reply tests/pytorch/mock_sparse2/test_reduction.py:9]: I'll try if it works on Windows first and then decide (now that I upgraded Windows CI to PyTorch 1.12)

File: python/dgl/mock_sparse2/reduction.py (hunk 17 above)
Comment (BarclayII) [reply python/dgl/mock_sparse2/reduction.py:9]: It's in `mock_sparse` so I added it.

File: tests/pytorch/mock_sparse2/test_reduction.py (hunk 16 above)
Comment (BarclayII) [reply tests/pytorch/mock_sparse2/test_reduction.py:117]: Done.

File: tests/pytorch/mock_sparse2/test_reduction.py (hunk 15 above)
Comment (BarclayII) [reply tests/pytorch/mock_sparse2/test_reduction.py:1]: I used `doctest` module to cover all docstrings automatically.  We actually should do it for other modules as well.

File: tests/pytorch/mock_sparse2/test_reduction.py (hunk 12 above)
Comment (BarclayII) [reply tests/pytorch/mock_sparse2/test_reduction.py:10]: Yeah it seems that DGL sparse still has trouble compiling on Windows.  I think it's a CMake problem.  I'll fix it in later PRs.

**Review by mufeili — APPROVED** [review]

**Pull request merged**